File: generated_code/GPT4/Python/039_SplayTree.py
class SplayTree:
    class Node:
        __slots__ = ('key', 'left', 'right', 'parent')

        def __init__(self, key, parent=None):
            self.key = key
            self.left = None
            self.right = None
            self.parent = parent

    def __init__(self):
        self.root = None

    def _rotate_left(self, x):
        y = x.right
        if y is None:
            return
        x.right = y.left
        if y.left:
            y.left.parent = x
        y.parent = x.parent
        if x.parent is None:
            self.root = y
        else:
            if x == x.parent.left:
                x.parent.left = y
            else:
                x.parent.right = y
        y.left = x
        x.parent = y

    def _rotate_right(self, x):
        y = x.left
        if y is None:
            return
        x.left = y.right
        if y.right:
            y.right.parent = x
        y.parent = x.parent
        if x.parent is None:
            self.root = y
        else:
            if x == x.parent.left:
                x.parent.left = y
            else:
                x.parent.right = y
        y.right = x
        x.parent = y

    def _splay(self, x):
        while x.parent:
            if x.parent.parent is None:
                # Zig step
                if x == x.parent.left:
                    self._rotate_right(x.parent)
                else:
                    self._rotate_left(x.parent)
            else:
                if x == x.parent.left and x.parent == x.parent.parent.left:
                    # Zig-zig step
                    self._rotate_right(x.parent.parent)
                    self._rotate_right(x.parent)
                elif x == x.parent.right and x.parent == x.parent.parent.right:
                    # Zig-zig step
                    self._rotate_left(x.parent.parent)
                    self._rotate_left(x.parent)
                elif x == x.parent.right and x.parent == x.parent.parent.left:
                    # Zig-zag step
                    self._rotate_left(x.parent)
                    self._rotate_right(x.parent)
                elif x == x.parent.left and x.parent == x.parent.parent.right:
                    # Zig-zag step
                    self._rotate_right(x.parent)
                    self._rotate_left(x.parent)

    def insert(self, key):
        if self.root is None:
            self.root = self.Node(key)
            return
        node = self.root
        while True:
            if key == node.key:
                self._splay(node)
                return  # key already in tree, do nothing
            elif key < node.key:
                if node.left is None:
                    node.left = self.Node(key, parent=node)
                    self._splay(node.left)
                    return
                node = node.left
            else:
                if node.right is None:
                    node.right = self.Node(key, parent=node)
                    self._splay(node.right)
                    return
                node = node.right

    def _subtree_max(self, node):
        while node.right:
            node = node.right
        return node

    def _replace(self, u, v):
        if u.parent is None:
            self.root = v
        else:
            if u == u.parent.left:
                u.parent.left = v
            else:
                u.parent.right = v
        if v:
            v.parent = u.parent

    def delete(self, key):
        node = self.root
        parent = None
        while node:
            if key == node.key:
                self._splay(node)
                break
            parent = node
            if key < node.key:
                node = node.left
            else:
                node = node.right
        else:
            # key not found, splay parent if exists
            if parent:
                self._splay(parent)
            return  # Key not in tree, do nothing

        # Now node is at root
        if node.left is None:
            self._replace(node, node.right)
        elif node.right is None:
            self._replace(node, node.left)
        else:
            # Find max in left subtree
            max_left = self._subtree_max(node.left)
            if max_left.parent != node:
                self._replace(max_left, max_left.left)
                max_left.left = node.left
                max_left.left.parent = max_left
            self._replace(node, max_left)
            max_left.right = node.right
            max_left.right.parent = max_left
            self._splay(max_left)

    def search(self, key):
        node = self.root
        parent = None
        while node:
            if key == node.key:
                self._splay(node)
                return True
            parent = node
            if key < node.key:
                node = node.left
            else:
                node = node.right
        # Not found: splay the last accessed parent
        if parent:
            self._splay(parent)
        return False

    # Optional: for debugging/testing
    def inorder(self):
        def _inorder(node):
            if node:
                yield from _inorder(node.left)
                yield node.key
                yield from _inorder(node.right)
        return list(_inorder(self.root))

File: generated_code/GPT4/Java/039_SplayTree.java
public class SplayTree {
    private class Node {
        int key;
        Node left, right, parent;

        Node(int key, Node parent) {
            this.key = key;
            this.parent = parent;
        }
    }

    private Node root;

    // Public insert method
    public void insert(int key) {
        if (root == null) {
            root = new Node(key, null);
            return;
        }
        Node curr = root;
        Node parent = null;
        while (curr != null) {
            parent = curr;
            if (key < curr.key) {
                curr = curr.left;
            } else if (key > curr.key) {
                curr = curr.right;
            } else {
                splay(curr); // Key already exists, splay it
                return;
            }
        }
        Node newNode = new Node(key, parent);
        if (key < parent.key) {
            parent.left = newNode;
        } else {
            parent.right = newNode;
        }
        splay(newNode);
    }

    // Public delete method
    public void delete(int key) {
        Node node = findNode(key);
        if (node == null) {
            return; // Key not found, do nothing
        }
        splay(node);
        // Now, node is root
        if (node.left != null) {
            Node maxLeft = subtreeMax(node.left);
            splay(maxLeft); // Splay max of left subtree to root
            maxLeft.right = node.right;
            if (node.right != null) {
                node.right.parent = maxLeft;
            }
            root = maxLeft;
            maxLeft.parent = null;
        } else if (node.right != null) {
            root = node.right;
            root.parent = null;
        } else {
            root = null;
        }
        // node is now disconnected and eligible for GC
    }

    // Public search method: returns Integer or null
    public Integer search(int key) {
        Node node = findNode(key);
        if (node != null) {
            splay(node);
            return node.key;
        } else {
            return null;
        }
    }

    // Helper: find node with given key (does not splay)
    private Node findNode(int key) {
        Node curr = root;
        while (curr != null) {
            if (key < curr.key) {
                curr = curr.left;
            } else if (key > curr.key) {
                curr = curr.right;
            } else {
                return curr;
            }
        }
        return null;
    }

    // Helper: Splay node to root
    private void splay(Node x) {
        if (x == null) return;
        while (x.parent != null) {
            Node p = x.parent;
            Node gp = p.parent;
            if (gp == null) {
                // Zig
                if (x == p.left) {
                    rightRotate(p);
                } else {
                    leftRotate(p);
                }
            } else if (x == p.left && p == gp.left) {
                // Zig-zig
                rightRotate(gp);
                rightRotate(p);
            } else if (x == p.right && p == gp.right) {
                // Zig-zig
                leftRotate(gp);
                leftRotate(p);
            } else if (x == p.right && p == gp.left) {
                // Zig-zag
                leftRotate(p);
                rightRotate(gp);
            } else if (x == p.left && p == gp.right) {
                // Zig-zag
                rightRotate(p);
                leftRotate(gp);
            }
        }
        root = x;
    }

    // Helper: left rotate
    private void leftRotate(Node x) {
        Node y = x.right;
        if (y == null) return;
        x.right = y.left;
        if (y.left != null) {
            y.left.parent = x;
        }
        y.parent = x.parent;
        if (x.parent == null) {
            root = y;
        } else if (x == x.parent.left) {
            x.parent.left = y;
        } else {
            x.parent.right = y;
        }
        y.left = x;
        x.parent = y;
    }

    // Helper: right rotate
    private void rightRotate(Node x) {
        Node y = x.left;
        if (y == null) return;
        x.left = y.right;
        if (y.right != null) {
            y.right.parent = x;
        }
        y.parent = x.parent;
        if (x.parent == null) {
            root = y;
        } else if (x == x.parent.left) {
            x.parent.left = y;
        } else {
            x.parent.right = y;
        }
        y.right = x;
        x.parent = y;
    }

    // Helper: Max node in subtree
    private Node subtreeMax(Node x) {
        while (x.right != null) {
            x = x.right;
        }
        return x;
    }

    // Optional: For testing, inorder traversal to print the tree
    public void printInOrder() {
        printInOrder(root);
        System.out.println();
    }

    private void printInOrder(Node node) {
        if (node == null) return;
        printInOrder(node.left);
        System.out.print(node.key + " ");
        printInOrder(node.right);
    }
}